File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    let ans = 0;
    const stack = []; // [ [started_index - hegiht] ]
    heights.forEach(function(height,i){
        if(!stack.length || height >= stack.at(-1)[1]) stack.push([i,height]);
        else{
            let [temp, start_index, pre_height] = [i,stack.at(-1)[0],stack.at(-1)[1]];
            while(stack.length && height < pre_height){
                stack.pop();
                ans = Math.max(ans, pre_height*(i-start_index));
                temp = start_index;
                if(stack.length){
                    [start_index,pre_height]  = [stack.at(-1)[0],stack.at(-1)[1]];
                }
            }
            stack.push([temp,height]);
        }
    })
    while(stack.length){
        ans = Math.max(ans,stack.at(-1)[1]*(heights.length-stack.at(-1)[0]));
        stack.pop();
    } 
    return ans;
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int,int>> s; // index - height
        int max_area = 0;
        int n = heights.size();
        for(int i = 0; i<n; i++){
            if(s.empty() || heights[i] >= s.top().second) s.push({i,heights[i]});
            else{
                int temp = i;
                int pre_index = s.top().first;
                int pre_height = s.top().second;
                while(!s.empty() && heights[i] < pre_height){
                    s.pop();
                    max_area = max(max_area, pre_height*(i-pre_index));
                    temp = pre_index;
                    if(!s.empty()){
                        pre_index = s.top().first;
                        pre_height = s.top().second;
                    }
                }
                s.push({temp, heights[i]});
            }
        }
        while(!s.empty()){
            int curr_index = s.top().first;
            int curr_height = s.top().second;
            s.pop();
            max_area = max(max_area, curr_height * (n-curr_index));
        }
    return max_area;
    }
};